retrieval/scrape.go

```diff
@@ -0,0 +1,294 @@
+	reportAppender storage.SampleAppender
+
+	// A function to calculate an initial offset for a given
+	// scrape interval.
+	offset func(time.Duration) time.Duration
+
+	stopc, done chan struct{}
+	cancel      func()
+	mtx         sync.RWMutex
+}
+
+func newScrapeLoop(app storage.SampleAppender, target *Target, state *TargetStatus) loop {
+	return &scrapeLoop{
+		scraper:        target,
+		appender:       target.wrapAppender(app, true),
+		reportAppender: target.wrapAppender(app, false),
+		state:          state,
+		offset:         target.offset,
+	}
+}
+
+func (sl *scrapeLoop) active() bool {
+	return sl.cancel != nil
+}
+
+func (sl *scrapeLoop) run(ctx context.Context, interval, timeout time.Duration, errc chan<- error) {
+	if sl.active() {
+		return
+	}
+
+	// If no offset function is provided, default to starting immediately.
+	if sl.offset == nil {
+		sl.offset = func(time.Duration) time.Duration { return 0 }
+	}
+
+	ctx, sl.cancel = context.WithCancel(ctx)
+
+	sl.stopc = make(chan struct{})
+	sl.done = make(chan struct{})
+
+	defer func() {
+		close(sl.done)
+		sl.cancel = nil
+	}()
+
+	select {
+	case <-time.After(sl.offset(interval)):
+		// Continue after a scraping offset.
+	case <-ctx.Done():
+		return
+	case <-sl.stopc:
+		return
+	}
+
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+
+	for {
+		start := time.Now()
+
+		err := sl.scrape(ctx, timeout)
+		if err != nil {
+			log.Debugf("Scraping error: %s", err)
+		}
+
+		sl.report(start, time.Since(start), err)
+		if errc != nil {
+			errc <- err
+		}
+
+		// First check without the ticker to not initiate
+		// another scrape if termination was requested.
+		select {
+		case <-sl.stopc:
+			return
+		case <-ctx.Done():
+			return
+		default:
+		}
+
+		select {
+		case <-sl.stopc:
+			return
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+		}
+	}
+}
+
+// kill terminates the scraping loop immediately.
+func (sl *scrapeLoop) kill() {
+	if !sl.active() {
+		return
+	}
+	sl.cancel()
+}
+
+// stop finishes any pending scrapes and then kills the loop.
+func (sl *scrapeLoop) stop() {
+	if !sl.active() {
+		return
+	}
+	close(sl.stopc)
+	<-sl.done
+
+	sl.kill()
+}
+
+// scrape executes a single Scrape and appends the retrieved samples.
+func (sl *scrapeLoop) scrape(ctx context.Context, timeout time.Duration) error {
+	ch := make(chan *model.Sample, ingestedSamplesCap)
+
+	// Receive and append all the samples retrieved by the scraper.
+	go func() {
+		var (
+			smpl *model.Sample
+			ok   bool
+		)
+		for {
+			select {
+			case smpl, ok = <-ch:
+				if !ok {
+					return
+				}
+			case <-ctx.Done():
+				return
+			}
+			// TODO(fabxc): Change the SampleAppender interface to return an error
+			// so we can proceed based on the status and don't leak goroutines trying
+			// to append a single sample after dropping all the other ones.
+			sl.appender.Append(smpl)
+		}
+	}()
+
+	ctx, _ = context.WithTimeout(ctx, timeout)
+
+	return sl.scraper.Scrape(ctx, ch)
```

Comment:
We might want to avoid passing every single sample through a channel here. We did that once in the past (where admittedly channel sends were even much slower), but then noticed that it slowed everything to a crawl, then switched to sending batches of samples through channels. The current HEAD does just synchronous calls, I think.

Just benchmarking this it's 96ns per sample (with channel cap of 256) for the channel send vs. 0.31ns when just passing it to a function. For 800k samples/s, that'd be 77ms per second used on the channel send here. Can we just pass in the appender to `Scrape()` and then do something like this there:

```
		for _, smpl := range samples {
			select {
			case <-ctx.Done():
				return ctx.Err()
			default:
				app.Append(smpl)
			}
 		}
```